### ISSUE TITLE
`rextendr::document` succeeds, but `devtools::install` fails

### ISSUE THREAD

**Issue by bicarlsen** [extendr/extendr#532]
When running `rextendr::document` in RStudio, my package builds and is usable from the local environment, but unavailable to external scripts. Trying to install with `devtools::install` to make it accessible results in the the error
```
error: failed to get `my-local-dependency` as a dependency of package `mypackage v0.0.1 (/tmp/Rtmpj00J85/file121d35b3ecd69/mypackage.Rcheck/00_pkg_src/thot/src/rust)`

Caused by:
  failed to load source for dependency `my-local-dependency`

Caused by:
  Unable to update /tmp/Rtmpj00J85/file121d35b3ecd69/mypackage.Rcheck/core

Caused by:
  failed to read `/tmp/Rtmpj00J85/file121d35b3ecd69/mypackage.Rcheck/core/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
make: *** [Makevars:16: rust/target/release/libmypackage.a] Error 101
```
where `mypackage` is the name of the `rextendr` package, and `my-local-dependency` is another rust crate in the same local workspace that `mypackage` is in.

**Comment by Ilia-Kosenkov** [comment]
We do not support multiple crates in one R package. It may work but we have never verified that. What you see is the path-related error (I suspect you use relative paths to point to another crate).

We need a reprex repository to play with.

**Comment by bicarlsen** [comment]
Yes, the `mypackage` crate uses a `path` to refer to the `my-local-dependency` crate.

I can create an MRE repo later today if you want. Where does this build process occur? Maybe I can start playing around with it too.

**Comment by Ilia-Kosenkov** [comment]
The package is built by R and we have little control over it. What we do is provide Makevars files where we compile rust library and then produce the final dll which is then added to the R package.

**Comment by bicarlsen** [comment]
Where does the DLL get produced?

**Comment by Ilia-Kosenkov** [comment]
Check package/src/Makevars

**Comment by CGMossa** [comment]
An MRE could be good, but this stuff is a giant headache and either it can
be solved by manually loading libraries in the package loader or messy,
messier things...

I'd just drop it if I were you unless it is deeply important to you.

**Comment by bicarlsen** [comment]
Unfortunately, it is deeply important to me.

**Comment by CGMossa** [comment]
Right, no worries. Just explain the full context and we'll give it a go.

**Comment by bicarlsen** [comment]
Here is an MRE to play with: [`extendr` multiple crates MRE](https://github.com/bicarlsen/extendr_multiple_crates-mre).
Please let me know if you have any questions.

**Comment by CGMossa** [comment]
Alright, so I have an answer for that.
I'm sorry that the structure will not resemble yours particularly, but I think this is the way to go.

First, conceptually the R-project is the main project. The rust crate(s) are secondary to it.
Thus I would expect that the repo would be an r project repo first, and rust
package(s) (which are a collection of crates) are contained within `src/rust/<here>`. 

In that in mind, a workspace be constructed within `src/rust/<workspace>`. 

But why would you want a workspace? Maybe you've got a project e.g. a new tool to
provide documentation for r-projects. Here you'd want a workspace to represent this tool. Say we call it `rdoctool`. Then that would be `src/rust/rdoctool`.

In there you might want a standalone crate for take r source code and producing `JSON` of some sort. I don't know, this is an example.
You can create that with `cargo new --lib rdocjson`.
And maybe next to it you'd have the crate that defines the r-api for this `rdoctool`. I'd name that package `rapi`.

The only thing this workspace has to do, is of course ensure that there is a library produced, and wrapper code of course, that the r-side has to take in.
Thus the `rapi` in `src/rust/rdoctool/rapi` has a `Cargo.toml` file like this:

```toml
[package]
name = "rapi"
version.workspace = true
edition.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[lib]
crate-type = ['staticlib']
name = 'rdoctool'

[dependencies]
extendr-api = '*'
rdocjson = { path = "../rdocjson" }
```

I've made this example and ran it and it works. 
I've included a comment or two in the commits, so you'll have to read commit messages to get the process of creating it.
But I've tested it and it works.

https://github.com/CGMossa/rdoctool

Let me know if this works for you.
In the mean time, please rename the issue here to something that would reflect this question.

**Comment by bicarlsen** [comment]
Okay, so this issue is fixed if you use absolute paths to reference local crates. It appears that because the cargo build step from `Makevars` is occurring in a temporary directory relative paths are being interpreted from that directory instead of the original directory.

**Comment by CGMossa** [comment]
Not only that, I also changed the Makevars to include `-p rapi` so that it would specifically build just the `crate` with the R-api in mind.
Further, the name of the produced lib has to match what R is expecting. So the `rapi` crate is the one producing the `rdoctool`-lib, and not the "workspace" itself.

**Comment by bicarlsen** [comment]
For me, setting the `-p` flag with relative paths for the local dependencies resulted in the original error.

**Comment by Ilia-Kosenkov** [comment]
@bicarlsen , if you could strip down your project, preserving only the structure of your R package/rust crates and upload it as a github repo, this would definitely speed up our investigation.
I must say that using absolute paths is not transferrable -- it will only work on your machine, while the goal is (I presume) to be able to distribute this package to others.

**Comment by bicarlsen** [comment]
> Here is an MRE to play with: [`extendr` multiple crates MRE](https://github.com/bicarlsen/extendr_multiple_crates-mre). Please let me know if you have any questions.

@Ilia-Kosenkov I posted an MRE here, but let me know if you had something else in mind.

The reason I noted that absolute paths worked is because when running `devtools::install()` for the `myextendr` package with the `my-extendr-lib` as a dependency I receive the error
```
error: failed to get `my-extendr-lib` as a dependency of package `myextendr v0.1.0 (/tmp/RtmpycOShC/R.INSTALL25232e6988fc/myextendr/src/rust)`

Caused by:
  failed to load source for dependency `my-extendr-lib`

Caused by:
  Unable to update /tmp/RtmpycOShC/R.INSTALL25232e6988fc/my-extendr-lib

Caused by:
  failed to read `/tmp/RtmpycOShC/R.INSTALL25232e6988fc/my-extendr-lib/Cargo.toml`

Caused by:
  No such file or directory (os error 2)
make: *** [Makevars:16: rust/target/release/libmyextendr.a] Error 101
ERROR: compilation failed for package ‘myextendr’
* removing ‘/home/me/R/x86_64-pc-linux-gnu-library/4.2/myextendr’
Error in `(function (command = NULL, args = character(), error_on_status = TRUE, …`:
! System command 'R' failed
```
Indicating that the `cargo build` was taking place in a `/tmp` directory and the relative path of the `my-extendr-lib` was referencing this `/tmp` dir instead of the original directory where the project is actually located. Thus, by setting the path of the `my-extendr-lib` dependency to be absolute the issue was resolved.

So it seems there isn't some more fundamental underlying issue, it's just that the correct path resolution needs to take place during the `devtools::install` process. It seems that this could be resolved with some changes to the `Makevars` file, but I'm not familiar enough with the system to offer suggestions as to what those changes would be.

**Comment by Ilia-Kosenkov** [comment]
Sorry I missed that. Yep, I see the issue here. You develop the project as if it is Rust-first. Instead, you need to approach it as R-first package with elements of Rust. R package should be at the top level, and all things rust-related should go into `/src/rust` subfolder.

**Comment by bicarlsen** [comment]
@Ilia-Kosenkov, in my actual project I have a fairly large set of local crates that the my R package depends on, so it is impractical to put have to copy them all into the R package and maintain two sets of them. Do you think another solution is possible allowing the use of relative paths outside the R package?

**Comment by Ilia-Kosenkov** [comment]
How do you plan to use your R package? Is it intended to work only on your local machine? If yes, then you can hard code absolute paths and forget about it. 
If however, you want to distribute it, then the best approach is probably to have your rust-only code pushed to some repository/crates.io and then in you R package you still only have Rust code necessary for interop with R and a dependency on your Rust crate with all of the logic.

The bottom line is, separate your R package from your complex Rust project.

**Comment by JosiahParry** [comment]
Note that this is possible and it uses {tomledit} to do so. See https://github.com/JosiahParry/anime/blob/main/r/bootstrap.R
